torch/distributed/_functional_collectives.py

```diff
@@ -226,17 +226,20 @@ def _expand_group(group: RANK_TYPES, tag: str = "") -> Tuple[str, List[int], int
         group_size = len(rankset)
         tag = tag or c10d._get_group_tag(group)
     elif isinstance(group, dt.DeviceMesh):
-        rankset = group.mesh.flatten().tolist()
-        group_size = group.mesh.size(0)
-        rankset = group.mesh.swapdims(-1, 0).reshape(-1, group_size).flatten().tolist()
-        tag = tag or c10d._get_group_tag(group.get_dim_groups()[0])
+        assert group.ndim == 1, "Only 1D mesh is supported, pass in (DeviceMesh, int) together if mesh > 1D"
+        # TODO: it should run collective in the whole mesh instead of dim 0
+        mesh_pg = group.get_dim_groups()[0]
+        rankset = dist.get_process_group_ranks(mesh_pg)
+        group_size = len(rankset)
+        tag = tag or c10d._get_group_tag(mesh_pg)
     elif isinstance(group, tuple):
         if len(group) == 2 and isinstance(group[0], dt.DeviceMesh) and isinstance(group[1], int):
             dmesh = group[0]
             dim = group[1]
-            group_size = dmesh.mesh.size(dim)
-            rankset = dmesh.mesh.swapdims(-1, dim).reshape(-1, group_size).flatten().tolist()
-            tag = tag or c10d._get_group_tag(dmesh.get_dim_groups()[dim])
+            dim_group = dmesh.get_dim_groups()[dim]
+            rankset = dist.get_process_group_ranks(dim_group)
```

Comment:
This prevents us from later doing collective algebraic fusion as the compiler won't see the whole mesh.

Reply:
Yeah we won't see the whole mesh, but given that this branch is taking a tuple of mesh, mesh_dim, it should be the correct behavior. 


filed an separate issue to track the whole mesh https://github.com/pytorch/pytorch/issues/97991.